### PR TITLE
Refactor aggregation base classes to remove doEquals() and doHashCode()

### PR DESCRIPTION
**Pull request by csoulios**
A number of the aggregation base classes have an abstract `doEquals()` and `doHashCode()` (e.g. `InternalAggregation.java`, `AbstractPipelineAggregationBuilder.java`).

Theoretically this is so the sub-classes can add to the `equals`/`hashCode` and don't need to worry about calling `super.equals()`.  In practice, it's mostly just confusing/inconsistent.  And if there are more than two levels, we end up with situations like `InternalMappedSignificantTerms` which has to call `super.doEquals()` which defeats the point of having these overridable methods.

This PR removes the `do` versions and just use `equals`/`hashCode` ensuring the super when necessary.

This PR is part of #41713 refactoring meta

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/pipeline/AbstractPipelineAggregationBuilder.java [hunk 1]
```diff
@@ -171,32 +171,20 @@ protected boolean overrideBucketsPath() {
...
-        if (!Objects.equals(name, other.name))
-            return false;
-        if (!Objects.equals(type, other.type))
-            return false;
-        if (!Objects.deepEquals(bucketsPaths, other.bucketsPaths))
-            return false;
-        if (!Objects.equals(metaData, other.metaData))
-            return false;
-        return doEquals(obj);
+        return Objects.equals(type, other.type)
```
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/search/aggregations/pipeline/AbstractPipelineAggregationBuilder.java]: ++ for cleaning this up!

File: server/src/main/java/org/elasticsearch/search/aggregations/pipeline/BucketMetricsPipelineAggregationBuilder.java [hunk 2]
```diff
@@ -147,21 +147,22 @@ protected final XContentBuilder internalXContent(XContentBuilder builder, Params
...
         @SuppressWarnings("unchecked")
         BucketMetricsPipelineAggregationBuilder<AF> other = (BucketMetricsPipelineAggregationBuilder<AF>) obj;
         return Objects.equals(format, other.format)
-                && Objects.equals(gapPolicy, other.gapPolicy)
-                && innerEquals(other);
+            && Objects.equals(gapPolicy, other.gapPolicy);
     }
 
-    protected abstract boolean innerEquals(BucketMetricsPipelineAggregationBuilder<AF> other);
+//    protected abstract boolean innerEquals(BucketMetricsPipelineAggregationBuilder<AF> other);
```
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/search/aggregations/pipeline/BucketMetricsPipelineAggregationBuilder.java]: Think these are leftover?

File: server/src/test/java/org/elasticsearch/search/SearchModuleTests.java [hunk 3]
```diff
@@ -441,13 +441,21 @@ protected XContentBuilder internalXContent(XContentBuilder builder, Params param
             return null;
         }
 
-        @Override
-        protected int doHashCode() {
+//        @Override
```
Comment (polyfractal) [inline server/src/test/java/org/elasticsearch/search/SearchModuleTests.java]: Also looks leftover?

**Review by polyfractal — CHANGES_REQUESTED** [review]
Left a few minor notes, and we talked about the inline negation issue on slack.  Otherwise looks good! Should be mergeable after the negations are cleaned up and some leftover commented code.  Thanks!

File: server/src/main/java/org/elasticsearch/search/aggregations/pipeline/BucketMetricsPipelineAggregationBuilder.java (hunk 2 above)
Comment (csoulios) [reply server/src/main/java/org/elasticsearch/search/aggregations/pipeline/BucketMetricsPipelineAggregationBuilder.java]: Oops. Good catch. Fixed

File: server/src/test/java/org/elasticsearch/search/SearchModuleTests.java (hunk 3 above)
Comment (csoulios) [reply server/src/test/java/org/elasticsearch/search/SearchModuleTests.java]: Fixed

**Comment by not-napoleon** [comment]
This is great, thanks for taking it on @csoulios !

File: modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/support/ArrayValuesSourceAggregationBuilder.java [hunk 4]
```diff
@@ -354,30 +354,21 @@ public final XContentBuilder internalXContent(XContentBuilder builder, Params pa
...
-        if (!Objects.equals(valuesSourceType, other.valuesSourceType))
-            return false;
-        return innerEquals(obj);
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        if (obj == null || getClass() != obj.getClass()) return false;
+        if (super.equals(obj) == false) return false;
+        ArrayValuesSourceAggregationBuilder<?, ?> that = (ArrayValuesSourceAggregationBuilder<?, ?>) obj;
+        return valuesSourceType == that.valuesSourceType &&
+            targetValueType == that.targetValueType &&
```
Comment (polyfractal) [inline modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/support/ArrayValuesSourceAggregationBuilder.java]: Is there a reason for the mix of `Objects.equals()` vs `==` here?

File: server/src/main/java/org/elasticsearch/search/aggregations/metrics/PercentileRanksAggregationBuilder.java [hunk 5]
```diff
@@ -269,37 +269,40 @@ protected XContentBuilder doXContentBody(XContentBuilder builder, Params params)
     }
 
     @Override
-    protected boolean innerEquals(Object obj) {
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        if (obj == null || getClass() != obj.getClass()) return false;
+        if (super.equals(obj) == false) return false;
         PercentileRanksAggregationBuilder other = (PercentileRanksAggregationBuilder) obj;
-        if (!Objects.equals(method, other.method)) {
```
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/search/aggregations/metrics/PercentileRanksAggregationBuilder.java]: :+1:

File: server/src/main/java/org/elasticsearch/search/aggregations/metrics/PercentilesAggregationBuilder.java [hunk 6]
```diff
@@ -292,35 +292,39 @@ protected XContentBuilder doXContentBody(XContentBuilder builder, Params params)
...
 
     @Override
-    protected boolean innerEquals(Object obj) {
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        if (obj == null || getClass() != obj.getClass()) return false;
+        if (super.equals(obj) == false) return false;
+
         PercentilesAggregationBuilder other = (PercentilesAggregationBuilder) obj;
         if (!Objects.equals(method, other.method)) {
```
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/search/aggregations/metrics/PercentilesAggregationBuilder.java]: While we're making changes, would you mind fixing this `!Objects.equals(...)` too?  Thanks!

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceAggregationBuilder.java [hunk 7]
```diff
@@ -344,34 +344,24 @@ public final XContentBuilder internalXContent(XContentBuilder builder, Params pa
...
-        if (!Objects.equals(targetValueType, other.targetValueType))
-            return false;
-        if (!Objects.equals(timeZone, other.timeZone))
-            return false;
-        if (!Objects.equals(valueType, other.valueType))
-            return false;
-        if (!Objects.equals(valuesSourceType, other.valuesSourceType))
-            return false;
-        return innerEquals(obj);
+        return valuesSourceType == other.valuesSourceType &&
```
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceAggregationBuilder.java]: `Objects.equals()` here?

**Review by polyfractal — APPROVED** [review]
Left a few more comments on the followup commit, but once those are addressed this LGTM!

Thanks @csoulios!  This ended up touching waaay more classes than I expected, appreciate the effort it took to work through all these :)

File: modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/support/ArrayValuesSourceAggregationBuilder.java (hunk 4 above)
Comment (csoulios) [reply modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/support/ArrayValuesSourceAggregationBuilder.java]: This was auto-generated by IntelliJ. I thought it doesn't hurt leaving it like this as I saw `==` notation used at some other classes already.

Comment (csoulios) [reply modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/support/ArrayValuesSourceAggregationBuilder.java]: Changing those to `Objects.equals()` as requested

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceAggregationBuilder.java (hunk 7 above)
Comment (csoulios) [reply server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceAggregationBuilder.java]: Fixed

**Pull request merged**